MCL

ettenhup opened the pull request:
Hi Benjamin,
I have now reimplemented MCL with the Eigen::SparseMatrix type and removed my own implementation of sparse matrices. I think this should also solve some performance issues. Could you please comment on the code as is for now as I think there is not much to do about the overall MCL procedure. Most critically, we need a good way to report clusters, and possibly improve on how data is piped from the aligner to the clustering and the command-line options should also be improved.

On another note: there is some commented code which should go into the test-framework, we should possibly talk about how to do that.

Cheers,
Patrick

bbuchfink commented on src/basic/config.cpp:
```diff
@@ -143,6 +143,10 @@ Config::Config(int argc, const char **argv, bool check_io)
 
+	Options_group cluster("Cluster options");
+	cluster.add()
+		("calgo", 0, "Clustering algorithm (0=multi-step/1=transitive-closure)", cluster_algo, -1);
```
A string option would be prefered (more intuitive for the user).

ettenhup replied:
I agree. I thought about three possibilities, let me know what you think/want.
1) a specific keyword for each clustering algorithm available, e.g. --MCL or --Superfast-fuzzy-clustering-yeah
2) repurposing the --algo option which takes a keyword argument "MCL" or "Superfast-fuzzy-clustering-yeah"
3) Similar to the above with an extra option (not --calgo but) --clustering-algo with the aforementioned keyword argument

I personally would tend to option 3 (also because I am not sure you may want to use both --algo and --clustering-algo in the same run). If you have a different opinion/idea, please let me know.

bbuchfink replied:
I would prefer option 3 as well.

bbuchfink commented on src/run/cluster.cpp:
```diff
@@ -16,6 +16,7 @@ You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ****/
 
```
This file should probably be moved to the /cluster/ directory.

ettenhup replied:
Done

bbuchfink commented on src/run/cluster.cpp:
```diff
@@ -151,10 +244,273 @@ void run() {
+	const size_t seq_count = db.ref_header.sequences;
+	statistics.reset();
+	config.command = Config::blastp;
+	config.no_self_hits = true;
```
The `no_self_hits` option probably needs to be reworked for self-alignment, as it relies on accession and alignment length, which could be broken by repeat masking (see align/culling.cpp:90).

ettenhup replied:
So it looks like setting it to false would be the preferred option here anyways. I think I just kept it because I was unsure as the numbers of clusters were so different (but we have resolved that in our email conversation). For MCL it is preferable to have the self-hits included as the graph needs to have these self-loops (and which I am currently adding). This will also enable more general similarity measures to be used for MCL (I don't like the idea of adding a "neutral" value as suggested in one paper).
At work, I have observed though that sometimes diamond does not find self-hits even if this option is enabled, but I don't know if that was a bug in the specific version of diamond we have been running for a while (and of course I have not checked since). I will enable this and throw a warning if a diagonal element is missing in the matrix.

bbuchfink replied:
Not finding a self hit should be either due to the repeat masking or due to the hit not being below the evalue cutoff. I'm not aware of any other way this could happen, but let me know if you find an example.

ettenhup replied:
Will do. Resolving this conversation for now.

bbuchfink commented on src/run/cluster.cpp:
```diff
@@ -60,6 +62,103 @@ struct Neighbors : public vector<vector<int>>, public Consumer {
+			components.emplace_back(move(component));
+			indices.emplace_back(move(order));
+		}
+		return std::make_pair(indices, components);
```
Also use std::move here?

ettenhup replied:
Done.

bbuchfink commented on src/run/cluster.cpp:
```diff
@@ -89,26 +188,22 @@ vector<int> cluster(DatabaseFile &db, const vector<bool> *filter) {
 	return Util::Algo::greedy_vortex_cover(nb);
 }
 
-void run() {
```
Probably should use a separate .cpp file for each clustering algorithm.

ettenhup replied:
Done. This change made it necessary to implement the command-line option as well. I have done this with option number 3 given above. If you want something else, please let me know.

bbuchfink commented on src/run/cluster.cpp:
```diff
@@ -151,10 +244,273 @@ void run() {
+	double diff_norm = std::numeric_limits<double>::max();
+	*m  = get_gamma(m,1); // This is to get a matrix of random walks on the graph -> TODO: find out if something else is more suitable
+	while( iteration < 100 && diff_norm > 1e-6*m->rows() ){
+		Eigen::SparseMatrix<float> msquared = ((*m) * (*m)).pruned(1.0, std::numeric_limits<float>::epsilon());
```
This could probably be optimized by avoiding some memory allocations for new matrices, see here: http://eigen.tuxfamily.org/index.php?title=FAQ#Where_in_my_program_are_temporary_objects_created.3F

ettenhup replied:
Possibly. I am not sure how to supply Eigen with the required memory for the temporary object though. I mean the memory requirements throughout the iterations are pretty static and it would be nice to just reuse the allocated space (and I hoped that Eigen does that under the hood). Do you have any reference on how to do this (if not, I can try and google it myself ;) )

bbuchfink replied:
The idea is basically to avoid the creation of temporarys, something like this:

```
Eigen::SparseMatrix<float> squared, diff, update;
while(...) {
  squared = m;
  squared *= m;
  get_gamma(squared, update)
  diff = m;
  diff -= update;
}
```
Not sure if Eigen provides a better interface that would also allow to avoid copy operations, and how much of a performance difference this actually makes.

ettenhup replied:
The *= operator does not exist for the sparse matrix type, otherwise this is implemented.

bbuchfink commented on src/run/disjoint_set.h:
```diff
@@ -0,0 +1,216 @@
+
+template<typename T> class LazyDisjointIntegralSet : public LazyDisjointSet<T> {
+private:
+	vector<Node<T>*> nodes;
```
This would maybe be more efficient using a `vector<int>` to store the parent index for each sequence, to avoid memory allocations.

ettenhup replied:
I don't think this is the most expensive step right now. I will keep it in mind and return to it if it turns out to be a problem.

bbuchfink commented:
Hi Patrick,

MCL reports its clusters as a tsv where each line is a cluster and contains a list of cluster members. Another option would be a TSV where each line contains the query accession and a cluster index (I would prefer this format probably).

I agree that piping the alignment data should be done more efficiently. There is a simple binary format available (see `output/output_format.h:198`), you could modify this accordingly for your needs.

It would be great to also provide some unit tests for this. The test framework can only run the search workflow so far. So the `TestCase` struct would have to be extended to also contain a workflow function, which would be `Workflow::Cluster::run` in your case. This function should then also take a `Workflow::Search::Options` parameter, which is used by the test framework to pass input/output files as a parameter instead of using the command line. The `TestCase` structure may also need to be extended to support different input files, right now only one small fasta file is always used.

Best regards,

Benjamin

ettenhup commented:
> Another option would be a TSV where each line contains the query accession and a cluster index (I would prefer this format probably).

I think this is a good choice as it resembles the BLAST output. Do you expect the accessions to come in order? It is a bit simpler to print the output in the order of the clusters rather than the order of the original sequences in the fasta.

bbuchfink commented:
> 
> 
> > Another option would be a TSV where each line contains the query accession and a cluster index (I would prefer this format probably).
> 
> I think this is a good choice as it resembles the BLAST output. Do you expect the accessions to come in order? It is a bit simpler to print the output in the order of the clusters rather than the order of the original sequences in the fasta.

In the order of the clusters would be fine.

ettenhup commented:
Status: I have now implemented all the suggestions except for the tests.

bbuchfink commented:
Thanks for the update, I'll see to look over it. As for the tests, they are not a priority and you can also wait until I find the time to extend the test framework for other workflows.

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,443 @@
+	sparse_gamma_time += (double) std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - t).count() / 1000.0;
+}
+
+void MCL::get_gamma(Eigen::MatrixXf* in, Eigen::MatrixXf* out, float r){
```
One may consider using Eigen's builtin functions here instead, such as the pow() method of the matrix, as well as the colwise() method, see here: https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html

This could be faster due to the SIMD acceleration but I'm not sure if it would make a big difference here, but could be worth a try.

ettenhup replied:
I have considered using pow, but it is not supported, therefore I have implemented the exp method. The reason for why it is not supported is that the matrix exponential is ill-defined if the eigenvector matrix is not invertible. To circumvent that I have now restricted the input to whole numbers. An alternative approach would be to _always_ use a symmetric similarity matrix, the the eigenvector matrix is unitary and only simple transposition is required to get the inverse. I will go with restricting the input to integral values for now, but if you have any preferences, please let me know.

ettenhup replied:
As for the colwise, as far as I can see, I would need two calls to colwise (reduction and broadcasting), while iterating through will only give me a single pass through the data. I can implement a version for performance testing.

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,443 @@
+	config.command = Config::blastp;
+	config.no_self_hits = false;
+	config.output_format = { "clus" };
+	config.query_cover = 80;
```
This should be configurable by the user.

ettenhup replied:
Which line does this comment apply to (this is possibly due to my github ignorance), 325? If so, I agree :)

bbuchfink replied:
Yes, I meant the query and subject cover.

ettenhup replied:
Thanks. Done

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,443 @@
+	config.algo = 0;
+	config.index_mode = 0;
+	config.freq_sd = 0;
+	config.max_alignments = numeric_limits<uint64_t>::max();
```
This should be configurable with a default value lower than everything, for better performance. There is a small discussion about this at https://micans.org/mcl/ (see Protocols). We should probably find some reasonable default value here, figuring out what works best will probably be left to the user.

ettenhup replied:
good point, although I would assume that a clustering would end up with more clusters than if you consider all of the hits. As you said, it should be up to the user. I also thought about exposing the similarity metric in order to make it even more customizable

ettenhup replied:
:) see below

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,443 @@
+	uint32_t nComponents = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 0;});
+	uint32_t nComponentsLt1 = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 1;});
+	timer.finish();
+	cout << "Found " << nComponentsLt1 <<" ("<<nComponents<< " incl. singletons) disconnected components" << endl;
```
Writing to `cout` should be reserved for output. You can use the `message_stream` object here.

bbuchfink commented on src/output/output_format.cpp:
```diff
@@ -131,4 +133,19 @@ void Bin1_format::print_match(const Hsp_context& r, const Metadata &metadata, Te
+void Clustering_format::print_match(const Hsp_context& r, const Metadata &metadata, TextBuffer &out) {
+	if (r.query_id <= r.subject_id) {
+		out.write((uint32_t)r.subject_id);
+		// TODO: enable other clustering measures
```
Agreed, this should be configurable by the user. Other measures could be the bit score, the negative log of the evalue, or the bit score divided by the length of the longer sequence. It would also be interesting to have the Needleman-Wunsch global alignment score and sequence identity here, although Diamond can't compute this at the moment.

ettenhup replied:
would the Needleman-Wunsch ga-score also be stored in the Hsp_context structure (or something else that is accessible from the print_match function?

ettenhup replied:
all of them and more have now been implemented

bbuchfink commented on src/cluster/cluster.h:
```diff
@@ -0,0 +1,26 @@
+/****
+DIAMOND protein aligner
+Copyright (C) 2013-2018 Benjamin Buchfink <buchfink@gmail.com>
```
For files that contain mostly your own code, please put an appropriate copyright notice here, with you and/or your employer as the copyright holder.

ettenhup replied:
Will do, but what about files where we have both modified the code?

bbuchfink replied:
You should amend the copyright notice then, although for files that you only changed minimally you can also leave it, if that's ok with you.

ettenhup replied:
For now I have chosen to keep the header as-is and replace the years, name and email address with my (private) credentials. I will check back with Q, what they would prefer, but for the current work I think it would also be appropriate to put my own creds since I did this in my own time.
Do you otherwise agree on that I keep the header "DIAMOND protein aligner"? I think that would be reasonable, since the code is an integral part of DIAMOND and cannot be used as a standalone tool.

bbuchfink replied:
Yes, I think that's fine.

ettenhup replied:
I have updated the headers according to the requirements of my employer and I have not put any shared copyright anywhere.

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,443 @@
+	}
+	else {
+		// TODO: check when/if using dense matrices for non-sparse components improves performance
+		for(uint32_t iComponent = 0; iComponent<indices.size(); iComponent++){
```
Should this loop be multithreaded?

ettenhup replied:
I think that would help quite a bit. I will look into it.

ettenhup replied:
I have now multithreaded this part.

ettenhup commented:
Hi Benjamin,
I have now addressed the remaining issues, please have a look and tell me what you think. Specifically, I have implemented:
- A more general handling of the clustering metric, i.e. anyone can build their own metric using the variable names specified by diamond, i.e. one can for example specify --cluster-similarity "exp(-evalue)*(qcovhsp+scovhsp)/2*pident" or similar. Supported functions are currently max, min, exp and log, but more could be added if necessary
- Multithreading has been added to the loop over disconnected components
- All the options you have noted, are now controlled through the config structure and not set anymore.

Again, I would appreciate your feedback.

ettenhup commented:
- Still need to look into the multithreading, does not seem to work as expected.
- I think the output should contain extra information about the attractors of MCL

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+#ifdef MCL_TIMINGS
+	chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
+#endif
+	vector<Eigen::Triplet<float>> data;
```
Maybe a reseve for this vector would make sense?

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+	// Note, we will access the clustering_result from several threads below and a vector does not guarantee thread-safety in these situations.
+	// Note also, that the use of the disjoint_set structure guarantees that each thread will access a different part of the clustering_result
+	uint32_t* clustering_result = new uint32_t[db->ref_header.sequences];
+	for(uint64_t iseq = 0; iseq < db->ref_header.sequences; iseq++){
```
Consider using std::fill

ettenhup replied:
good suggestion. done.

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+		uint32_t cluster_id = iThr;
+		//uint32_t my_counter = component_counter.fetch_add(1, memory_order_relaxed);
+		//while(my_counter < max_counter){
+		for(uint32_t my_counter = iThr; my_counter<max_counter; my_counter += nThreads){
```
The solution with `component_counter` seems to be better here.

ettenhup replied:
Yes, but the issue was that I have never observed a cpu usage over 100% even though the threads have been spawned. My assumption was that the fetch_add call was blocking (on the system I was running on - I didnt check the is_lock_free() function). But also with the changes here the cpu usage in the clustering phase was at a 100%, so I am a bit puzzled. Considering that for the test I have been running, the clustering is 100x faster than creating the adjacency matrix, this is probably not an issue (yet). But we may want to rethink whether the alignment phase is actually necessary for creating an (approximate) adjacency matrix

ettenhup replied:
I have just confirmed that is_lock_free is true. I am a bit puzzled what is going on. I will switch back to the solution with the atomic update but change the granularity, such that each of the threats takes a few iterations in one shot, eg. 100 or so (could be made dependent on the number of threats and total number of disconnected components).

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+				unordered_set<uint32_t> attractors;
+
+				//TODO: a size limit for the dense matrix should control this as well
+#ifdef MCL_TIMINGS
```
The `high_resolution_clock` was very fast in my tests, so you can also go without these ifdefs if you want.

ettenhup replied:
I have mostly removed this due to the assumption described above and to the fact that I am collecting the timings with atomic fetch_add updates

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+#endif
+				}
+				for(unordered_set<uint32_t> subset : list_of_sets){
+					assert(cluster_id < 0x00ffffff);
```
Do we support a max. of 16.7 million clusters?

ettenhup replied:
right now yes. This can be changed easily. One question is whether we want to be able to lable three or only two kinds of nodes, i.e. singletons (which may be inferred from the clusters), normal nodes and attractor nodes. Removing the singleton label gives a factor two, i.e. roughly 33 milion clusters, if that is not enough, we can just switch to a 64 bit integer instead (or bigger if necessary) at the cost of an increased memory footprint

bbuchfink replied:
Maybe use a struct or a second array for this? Seems a bit cleaner.

bbuchfink commented on src/cluster/mcl.cpp:
```diff
@@ -0,0 +1,570 @@
+	delete[] max_sparsities;
+
+	timer.finish();
+	cout << "Found "
```
`cout` should be reserved for the TSV output that goes into the file. You can use `message_stream` for this.

ettenhup replied:
done.

bbuchfink commented on src/output/clustering_variables.h:
```diff
@@ -0,0 +1,310 @@
+private:
+	VariableRegistry(){};
+	// To include new clustering algorithms add the instantiation here and in the cluster_registry.cpp file. Then add it to the StaticConstructor below
+	static QueryStart queryStart;
```
Length of query and subject could be useful.

ettenhup replied:
Good catch. I meant to add all of the defined variables in the header.
